api/all/src/main/java/io/opentelemetry/api/trace/SpanBuilder.kt

```diff
@@ -0,0 +1,371 @@
+ *
+ * <pre>`class MyClass {
+ * private final Tracer tracer;
+ *
+ * MyClass(OpenTelemetry openTelemetry) {
+ * tracer = openTelemetry.getTracer("com.anyco.rpc");
+ * }
+ *
+ * void doWork() {
+ * // Create a Span as a child of the current Span.
+ * Span span = tracer.spanBuilder("MyChildSpan").startSpan();
+ * try (Scope ss = span.makeCurrent()) {
+ * span.addEvent("my event");
+ * doSomeWork();  // Here the new span is in the current Context, so it can be used
+ * // implicitly anywhere down the stack.
+ * } finally {
+ * span.end();
+ * }
+ * }
+ * }
+`</pre> *
+ *
+ *
+ * There might be cases where you do not perform all the work inside one static scope and the
+ * Context is automatically propagated:
+ *
+ * <pre>`class MyRpcServerInterceptorListener implements RpcServerInterceptor.Listener {
+ * private final Tracer tracer;
+ * private Span mySpan;
+ *
+ * public MyRpcInterceptor(OpenTelemetry openTelemetry) {
+ * tracer = openTelemetry.getTracer("com.example.rpc");
+ * }
+ *
+ * public void onRequest(String rpcName, Metadata metadata) {
+ * // Create a Span as a child of the remote Span.
+ * mySpan = tracer.spanBuilder(rpcName)
+ * .setParent(extractContextFromMetadata(metadata)).startSpan();
+ * }
+ *
+ * public void onExecuteHandler(ServerCallHandler serverCallHandler) {
+ * try (Scope ws = mySpan.makeCurrent()) {
+ * Span.current().addEvent("Start rpc execution.");
+ * serverCallHandler.run();  // Here the new span is in the current Context, so it can be
+ * // used implicitly anywhere down the stack.
+ * }
+ * }
+ *
+ * // Called when the RPC is canceled and guaranteed onComplete will not be called.
+ * public void onCancel() {
+ * // IMPORTANT: DO NOT forget to ended the Span here as the work is done.
+ * mySpan.setStatus(StatusCode.ERROR);
+ * mySpan.end();
+ * }
+ *
+ * // Called when the RPC is done and guaranteed onCancel will not be called.
+ * public void onComplete(RpcStatus rpcStatus) {
+ * // IMPORTANT: DO NOT forget to ended the Span here as the work is done.
+ * mySpan.setStatus(rpcStatusToCanonicalTraceStatus(status);
+ * mySpan.end();
+ * }
+ * }
+`</pre> *
+ *
+ *
+ * This is a simple example where all the work is being done within a single scope and the
+ * Context is manually propagated:
+ *
+ * <pre>`class MyClass {
+ * private final Tracer tracer;
+ *
+ * MyClass(OpenTelemetry openTelemetry) {
+ * tracer = openTelemetry.getTracer("com.example.rpc");
+ * }
+ *
+ * void doWork(Span parent) {
+ * Span childSpan = tracer.spanBuilder("MyChildSpan")
+ * .setParent(Context.current().with(parent))
+ * .startSpan();
+ * childSpan.addEvent("my event");
+ * try {
+ * doSomeWork(childSpan); // Manually propagate the new span down the stack.
+ * } finally {
+ * // To make sure we end the span even in case of an exception.
+ * childSpan.end();  // Manually end the span.
+ * }
+ * }
+ * }
+`</pre> *
+ *
+ *
+ * see [SpanBuilder.startSpan] for usage examples.
+ */
+interface SpanBuilder {
```

Comment:
Open question: should DSL generate span w/ state, or should they be separate interfaces?